src/charm.py

```diff
@@ -213,6 +278,148 @@ def _on_get_rule_action(self, event: ActionEvent) -> None:
+        if not self._oathkeeper_service_is_running:
+            self.unit.status = WaitingStatus("Waiting for Oathkeeper service")
+            event.defer()
+            return
+
+        if not self._peers:
+            self.unit.status = WaitingStatus("Waiting for peer relation")
+            event.defer()
+            return
+
+        access_rules_locations = list()
+        for rule_type in ("allow", "deny"):
+            rules = self._render_access_rules(
+                rule_type=rule_type,
+                protected_urls=event.protected_urls,
+                allowed_endpoints=event.allowed_endpoints,
+                relation_app_name=event.relation_app_name,
+            )
+
+            if rules:
+                # Push the rules to the container
+                filename = f"{self._oathkeeper_access_rules_path}/access-rules-{event.relation_app_name}-{rule_type}.json"
+                self._container.push(filename, str(rules), make_dirs=True)
+                access_rules_locations.append(filename)
+
+        self._set_auth_proxy_relation_peer_data(
+            event.relation_id, dict(access_rules_locations=access_rules_locations)
+        )
+
+        try:
+            self._push_oathkeeper_config()
+        except Error as e:
+            logger.error(f"Failed to set new config: {e}")
+            self.unit.status = BlockedStatus("Failed to set new config, see logs")
+            self._pop_auth_proxy_relation_peer_data(event.relation_id)
+            return
+
+    def _rule_template(
+        self, rule_id: str, url: str, authenticator: str, mutator: str, error_handler: str
+    ) -> Dict:
+        return {
+            "id": rule_id,
+            "match": {"url": url, "methods": ["GET", "POST", "OPTION", "PUT", "PATCH", "DELETE"]},
+            "authenticators": [{"handler": authenticator}],
+            "mutators": [{"handler": mutator}],
+            "authorizer": {"handler": "allow"},
+            "errors": [{"handler": error_handler}],
+        }
+
+    def _render_access_rules(
+        self,
+        rule_type: str,
+        protected_urls: List[str],
+        allowed_endpoints: List[str],
+        relation_app_name: str,
+    ) -> Optional[List[Dict]]:
+        """Render access rules from a template."""
+        rules = list()
+
+        for url_index, url in enumerate(protected_urls):
+            # Parse url
+            if url.endswith("/"):
+                url = url[:-1]
+            # Replace with regex to match both http and https
+            url = url.replace("https", "<^(https|http)>")
+
+            if rule_type == "allow":
+                if not allowed_endpoints:
+                    return None
+                for endpoint in allowed_endpoints:
+                    allow_rule = self._rule_template(
+                        rule_id=f"{relation_app_name}:{endpoint}:{url_index}:allow",
+                        url=f"{url}/{endpoint}",
```

Comment:
```suggestion
                        url=f"^{url}/{endpoint}$",
```

Reply:
`^` is covered a few lines above in `url = url.replace("https", "<^(https|http)>")`. `$` is not needed because the rest of the allowed urls are exact matches, not regex.